Regression: libc++abi: terminating due to uncaught exception of type cv::Exception: OpenCV(4.9.0)

This file was able to be read by this tool in the past:
[testVertical.tiff.zip](https://github.com/user-attachments/files/18370035/testVertical.tiff.zip)


Exit code: -6
stdout: `[INFO] Importing image...`
stderr:
```
libc++abi: terminating due to uncaught exception of type cv::Exception: OpenCV(4.9.0) /Users/fractale/.conan2/p/b/openc9bbec52ee9964/b/src/modules/core/src/matrix_wrap.cpp:1300: error: (-2:Unspecified error) in function 'void cv::_OutputArray::create(int, const int *, int, int, bool, _OutputArray::DepthMask) const'
> Can't reallocate Mat with locked type (probably due to misused 'const' modifier) (expected: 'm.type() == CV_MAT_TYPE(mtype)'), where
>     'm.type()' is 0 (CV_8UC1)
> must be equal to
>     'CV_MAT_TYPE(mtype)' is 2 (CV_16UC1)
```

--- a/include/FileIO/FileIO.hpp
+++ b/include/FileIO/FileIO.hpp
@@ -15,8 +15,9 @@ class FileIO {
 	virtual ~FileIO() = default;
 
 	void setImg(cv::Mat newImage);
-	bool load();
 
+	bool loadImg();
+	bool loadLut();
 	bool saveImg(cv::Mat newImg) const;
 
 	[[nodiscard]] const cv::Mat_<cv::Vec3b>& getImg() const;
@@ -25,8 +26,6 @@ class FileIO {
 protected:
 	std::unique_ptr<CubeLUT> cube;
 	cv::Mat1b alphaChannel; // Used only for RGBA input images
-	bool loadImg();
-	bool loadLut();
 
 	virtual cv::Mat readImage(const std::string& inputPath) const;
 	virtual bool writeImage(const std::string &outputPath, cv::Mat newImg) const;

--- a/src/FileIO/FileIO.cpp
+++ b/src/FileIO/FileIO.cpp
@@ -10,12 +10,43 @@ FileIO::FileIO(const InputParams& params)
 	: inputPath(params.getInputImgPath()), outputPath(params.getOutputImgPath()), lutPath(params.getInputLutPath()),
 	  forceOverwrite(params.getForceOverwrite()), cube(std::make_unique<CubeLUT>()) {}
 
+namespace {
+uint cvDepthToInt(int depthCode) {
+	switch(depthCode) {
+		case CV_8U:
+		case CV_8S:
+			return 8;
+		case CV_16U:
+		case CV_16S:
+		case CV_16F:
+			return 16;
+		case CV_32S:
+		case CV_32F:
+			return 32;
+		case CV_64F:
+			return 64;
+	}
+	return 0;
+}
+
+cv::Mat normalizeImage(cv::Mat img) {
+	cv::Mat convertedMat;
+	cv::normalize(img, convertedMat, 0, 255, cv::NORM_MINMAX, CV_8U);
+	return convertedMat;
+}
+}
+
 bool FileIO::loadImg() {
 	std::cout << "[INFO] Importing image...\n";
 	auto sourceImg = readImage(inputPath);
 	if (sourceImg.empty()) {
 		std::cerr << fmt::format("[ERROR] Could not open input image file: {}\n", inputPath);
 		return false;
+	}
+
+	if (const auto depth = sourceImg.depth(); depth != CV_8U) {
+		std::cout << fmt::format("[WARN] Unsupported bit depth ({}) - output image will be converted to 8\n", cvDepthToInt(depth));
+		sourceImg = normalizeImage(sourceImg);
 	}
 
 	if (const auto channels = sourceImg.channels(); channels < 3 || channels > 4) {

--- a/src/FileIO/FileIO.cpp
+++ b/src/FileIO/FileIO.cpp
@@ -97,11 +97,6 @@ bool FileIO::loadLut()
 	return success;
 }
 
-bool FileIO::load()
-{
-	return loadImg() && loadLut();
-}
-
 const cv::Mat3b& FileIO::getImg() const
 {
 	return this->img;

--- a/src/TaskDispatcher/TaskDispatcher.cpp
+++ b/src/TaskDispatcher/TaskDispatcher.cpp
@@ -33,7 +33,14 @@ int TaskDispatcher::start()
 	}
 	FileIO fileIO{ parameters };
 
-	bool loadSuccessful = fileIO.load();
+	bool loadSuccessful{false};
+	try {
+		loadSuccessful = fileIO.loadImg() && fileIO.loadLut();
+	} catch (const std::exception& ex) {
+		std::cerr << fmt::format("[ERROR] Fatal exception: {}", ex.what());
+		return FAIL_EXIT;
+	}
+
 	if (!loadSuccessful) {
 		return FAIL_EXIT;
 	}

--- a/src/test/FileIOTest.cpp
+++ b/src/test/FileIOTest.cpp
@@ -151,3 +151,23 @@ TEST_F(FileIOTest, saveWithoutForceOverwrite) {
     EXPECT_CALL(loader, writeImage).WillOnce(Return(true));
     EXPECT_TRUE(loader.saveImg({}));
 }
+
+TEST_F(FileIOTest, readHighBitDepth) {
+    FileIOMock loader(params);
+    cv::Mat mat16(cv::Size(10, 10), CV_16UC3);
+    EXPECT_CALL(loader, readImage).WillOnce(Return(mat16));
+    EXPECT_TRUE(loader.loadImg());
+    const auto parsedMat16 = loader.getImg();
+    EXPECT_EQ(parsedMat16.depth(), CV_8U);
+    EXPECT_EQ(parsedMat16.channels(), 3);
+    EXPECT_EQ(parsedMat16.type(), CV_8UC3);
+
+    cv::Mat mat16alpha(cv::Size(10, 10), CV_16UC4);
+    EXPECT_CALL(loader, readImage).WillOnce(Return(mat16alpha));
+    EXPECT_TRUE(loader.loadImg());
+    const auto parsedMat16alpha = loader.getImg();
+    EXPECT_EQ(parsedMat16alpha.depth(), CV_8U);
+    EXPECT_EQ(parsedMat16alpha.channels(), 3);
+    EXPECT_EQ(parsedMat16alpha.type(), CV_8UC3);
+    EXPECT_FALSE(loader.getAlphaChannel().empty());
+}

